arch/sim/src/sim/up_i2cbuslinux.c

```diff
@@ -0,0 +1,310 @@
+ * Name: linux_i2cbus_transfer
+ *
+ * Description:
+ *   Provide i2c transfer
+ *
+ ****************************************************************************/
+
+static int linux_i2cbus_transfer(struct i2c_master_s *dev,
+                                 struct i2c_msg_s *msgs, int count)
+{
+  int ret = 0;
+  struct linux_i2cbus_master_s *priv = (struct linux_i2cbus_master_s *)dev;
+  int file = priv->file;
+  uint8_t *pack_buf = NULL;
+  struct i2c_rdwr_ioctl_data ioctl_data;
+  struct i2c_msg l_msgs[2];  /* We only support up to 2 messages */
+  ioctl_data.msgs = l_msgs;
+
+  /* Many i2c bus do not play well with combined messages via the Linux
+   * interface, this makes stitching things together a little harder
+   * because NuttX provides the ability to hold the bus without ending
+   * with a STOP which is not ideal in general, and not possible with
+   * Linux.
+   */
+
+  if ((msgs[0].flags & NUTTX_I2C_M_TEN) || (msgs[1].flags & NUTTX_I2C_M_TEN))
+    {
+      /* Linux also has somewhat poor support for 10bit addresses and they
+       * are quite rare so we just don't support them for now here
+       */
+
+      return -1;
+    }
+
+  ioctl_data.msgs = l_msgs;
+
+  if (count == 1)
+    {
+      if (msgs[0].flags & NUTTX_I2C_M_NOSTOP || \
+          msgs[0].flags & NUTTX_I2C_M_NOSTART)
+        {
+          /* Do not support leaving the bus hanging or try to send
+           * without first starting.
+           */
+
+          return -1;
+        }
+
+      l_msgs[0].addr = msgs[0].addr;
+      l_msgs[0].buf = msgs[0].buffer;
+      l_msgs[0].len = msgs[0].length;
+      l_msgs[0].flags = 0;
+      if (msgs[0].flags & NUTTX_I2C_M_READ)
+        {
+           l_msgs[0].flags |= I2C_M_RD;
+        }
+
+      ioctl_data.nmsgs = 1;
+    }
+  else if(count == 2)
+    {
+      /* Addresses should be the same */
+
+      if (msgs[0].addr != msgs[1].addr)
+        {
+          return -1;
+        }
+
+      /* Check if we are about to do a read of a register
+       * NuttX interface represents this as WRITE(NOSTOP) + READ
+       * Linux interface represents this as READ + READ
+       */
+
+      if (msgs[0].flags & NUTTX_I2C_M_NOSTOP && \
+          msgs[1].flags & NUTTX_I2C_M_READ)
+        {
+          l_msgs[0].addr = msgs[0].addr;
+          l_msgs[0].flags = 0;
+          l_msgs[0].buf = msgs[0].buffer;
+          l_msgs[0].len = msgs[0].length;
+
+          l_msgs[1].addr = msgs[1].addr;
+          l_msgs[1].flags = I2C_M_RD;
+          l_msgs[1].buf = msgs[1].buffer;
+          l_msgs[1].len = msgs[1].length;
+          ioctl_data.nmsgs = 2;
+        }
+      else if (!(msgs[0].flags & NUTTX_I2C_M_READ) && \
+               !(msgs[1].flags & NUTTX_I2C_M_READ) && \
+               (msgs[0].flags & NUTTX_I2C_M_NOSTOP) && \
+               (msgs[1].flags & NUTTX_I2C_M_NOSTART))
+        {
+          /* These writes are actually just a single write in Linux
```

Comment:
Linux i2c kernel driver can handle this case correctly, why we need do this manually?

Reply:
No it cannot for many of the bus drivers.

Reply:
Also it is very important to note that each message in Linux is defined by this
```
struct i2c_msg * msgs
    One or more messages to execute before STOP is issued to terminate the operation; each message begins with a START.
``
See how I check for `NUTTX_I2C_M_NOSTOP` that means that we MUST pack this into a single message or Linux will inject a STOP and START after the first message.

Reply:
Technically I could probably use NOSTART on the second message and meet the Linux interface spec but like I pointed out most drivers don't support doing that and I really want to makes sure we can use common USB ones otherwise this is not very useful.